apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriptionHandler.cs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License").
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at 
+//
+// https://www.apache.org/licenses/LICENSE-2.0 
+//
+// Unless required by applicable law or agreed to in writing, software 
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using static Google.Cloud.PubSub.V1.SubscriberClient;
+
+namespace Google.Cloud.PubSub.V1;
+
+/// <summary>
+/// This abstract class provides the message handler to handle and process the messages.
+/// It also provides the ACKnowledge (ACK) and Not ACKnowledge (NACK) response handlers which indicates the ACK/NACK status of each message for the subscription.
+/// The users of subscription should inherit from this class and provide their implementation to handle messages and their acknowledgement response.
+/// </summary>
+public abstract class SubscriptionHandler
+{
+    /// <summary>
+    /// The handler function that is passed all received messages.
+    /// This function may be called on multiple threads concurrently. Return <see cref="Reply.Ack"/> from this function
+    /// to ACKnowledge this message (implying it won't be received again); or return <see cref="Reply.Nack"/> to Not
+    /// ACKnowledge this message (implying it will be received again). If this function throws any Exception, then
+    /// it behaves as if it returned <see cref="Reply.Nack"/>.
+    /// </summary>
+    /// <remarks>
+    /// The users of subscription must implement this function to handle the messages.
+    /// </remarks>
+    public abstract Task<Reply> HandleMessage(PubsubMessage message, CancellationToken cancellationToken);
+
+    /// <summary>
+    /// This handler function is called when an ACKnowledge (ACK) operation returns, indicating the status of successful or permanently failed messages.
+    /// </summary>
+    /// <param name="responses">The read-only collection of the <see cref="AckNackResponse"/>
+    /// indicating the <see cref="SubscriberClient.Reply.Ack"/> status of successful or permanently failed messages. This collection may be empty but never null.
+    /// For exactly once delivery subscription, this collection contains only the successful or permanently failed messages. In this case,
+    /// the messages that have failed with temporary errors are not included in this collection as their status may change in the future.
+    /// For non-exactly once subscriptions, this collection contains the ACK status as success for all the messages. In this case,
+    /// a message may appear multiple times with the same status.</param>
```

Comment:
This is only if multiple acks were called for the same message, right? Please clarify this.

Reply:
For non exactly once subscription, as discussed in design doc, we are treating the ACK/NACK request as fire and forget for compatibility with other libraries, so we always send the Success status for ACK/NACK even if there may be a recoverable error like Deadline exceeded. In existing C# implementation, if there is a recoverable error like Deadline exceeded, the message is requeued for ACK/NACK. In such cases, the same message will appear in the ResponseHandler again and the status of request will be success. This will happen only in cases where the handler is used for non exactly once delivery subscriptions as we are not sharing the right status. In exactly once delivery subscription, we share the status of success and permanently failed messages only which are final, so it won't happen there.

Reply:
Acks/modacks should not be retried if EOS is disabled. Please change the behavior for non-EOS subscriptions. The value of retrying a fire-and-forget operation like ack/modack upon DeadlineExceeded is negligible.

So, for EOS and non-EOS subscriptions, each status returned should correspond to one user-initiated action.

Reply:
This is in `HandleAckResponse` method in `SubscriberClient` class (line 1690), which has not been modified as part of current EOD changes and has been existing this way in C# library. It checks if the error is recoverable, i.e. StatusCode of gRPC exception is one of DeadlineExceeded, ResourceExhausted, Unavailable, Cancelled, Aborted etc, it treats that error as recoverable and re-queues the ackId/nackId/extendId for non exactly once delivery. If this is not the right behavior, I think we can track this as an issue and fix that in a different PR, so that exactly once changes and non-exactly once changes can be differentiated and we would be able to recollect later why the change is done. (I can do that change immediately after this exactly once delivery PR) Please let me know if you have any concerns with this. 
@jskeet - Please share if you have insights about this behavior and your thoughts on handling this change.

Reply:
Yes, let's do this in a separate PR. You can copy this discussion into the issue for that one.

Reply:
Thanks Prad. I will create a new issue and refer this discussion.

Reply:
This issue is tracked in https://github.com/googleapis/google-cloud-dotnet/issues/8992. We can close this comment.